Firestore/core/src/firebase/firestore/local/local_serializer.h

```diff
@@ -107,6 +149,10 @@ class LocalSerializer {
   std::unique_ptr<model::NoDocument> DecodeNoDocument(
       nanopb::Reader* reader) const;
 
+  void EncodeQueryData(nanopb::Writer* writer,
+                       const QueryData& query_data) const;
+  QueryData DecodeQueryData(nanopb::Reader* reader) const;
```

Comment:
There are too many axes and permutations for each decoded type. We have public vs private, returning `StatusOr<T>` vs `T`, taking `Reader*` vs bytes and size vs `vector<uint8_t>`. It's not obvious to me which of these we need to write for the next decoded type and the amount of boilerplate here is crippling. Let's reduce the combinatorial space.

First: let's just standardize on all Decode methods just taking a Reader. Let's make it the caller's problem to figure out how to construct one. This is actually beneficial because it separates the concern of how to the Reader is constructed from this usage. Production code can use a helper function to directly create a Reader that wraps a grpc buffer/slice without having to pass bytes + size down to this layer.

The second part is the standard return type. Here I'm less sure what to do, but for a type Foo we should standardize on one of these:

```c++
Foo DecodeFoo(Reader* reader);
```
**Pro**
  * Composes nicely when one decode method calls another (which is almost all of them)
  * Fail early cases could be `if (!reader->status().ok()) return {};` (but this depends on a con below)

**Con**
  * All decoded types must be default constructible or have an `Invalid` helper
  * Caller must check `reader.status`. Could be mitigated with a `reader.StatusOr` helper (described below)

```c++
util::StatusOr<Foo> DecodeFoo(Reader* reader);
```
**Pro**
  * More directly encodes the possibility of failure
  * Types don't need `Foo::Invalid`

**Con**
  * Fail early cases look like `if (!reader->status().ok()) return reader->status();` which isn't so pretty
  * Duplicates status information; causes confusion about where to get it
  * Creating derived objects really stinks: `QueryData(std::move(query).ValueOrDie(), ...)`

```c++
absl::optional<Foo> DecodeFoo(Reader* reader);
```
**Pro**
  * Still directly encodes the possibility of failure
  * Types don't need `Foo::Invalid`
  * Status isn't duplicated.
  * Fail early cases look like `if (!reader->status().ok()) return nullopt;` which isn't bad.
  * Unifies with cases where a value is optionally present (though I don't know if this is really a thing)

**Con**
  * Constructing derived objects still stinks, but not as bad as StatusOr: `QueryData(*std::move(query), ...)`
  * Code bloat due to all the optional types

I think my least favorite of the bunch is returning `StatusOr<Foo>`, just because it makes composition of decode methods so noisy.

With some of the changes I suggested to avoid returning `Foo::Invalid` all over the place I could be OK with returning type `Foo` as well.

I find myself leaning toward `absl::optional<T>` but worry about code size. I wonder what the fully loaded cost of each optional instantiation is?

Reply:
(As I mentioned elsewhere) I'm not wild about returning invalid objects, so I'd lean toward statusor/optional. Having the status surfaced twice seem questionable, so I'm also leaning toward optional. (If code size becomes an issue, we can always fall back on invalid objects...)